multiqc/modules/biscuit/biscuit.py

```diff
@@ -0,0 +1,688 @@
+#!/usr/bin/env python
+
+""" MultiQC module to parse output from BISCUIT """
+
+from __future__ import print_function
+from collections import OrderedDict
+import logging
+import re
+
+from multiqc import config
+from multiqc.plots import beeswarm, linegraph, bargraph, table
+from multiqc.modules.base_module import BaseMultiqcModule
+
+# Initialise the logger
+log = logging.getLogger(__name__)
+
+def _uniquify_ID(data, header, namespace):
+    header2 = dict([(namespace+k, v) for k, v in header.items()])
+    data2 = {}
+    for sid, datum in data.items():
+        data2[sid] = dict([(namespace+k,v) for k, v in datum.items()])
+
+    return (data2, header2)
+
+# Log parsing regexes
+class MultiqcModule(BaseMultiqcModule):
+
+    def __init__(self):
+
+        # Initialise the parent object
+        super(MultiqcModule, self).__init__(name='BISCUIT', anchor='biscuit',
+        href="https://github.com/zwdzwd/biscuit",
+        info="is a tool to map bisulfite converted sequence reads and determine"\
+        " cytosine methylation states.")
+
+        # Set up data structures
+        self.mdata = {
+            'mapq': {},
+            'markdup': {},
+            'coverage': {},
+            'retention': {},
+        }
+
+        # Find and parse bismark alignment reports
+        for k in self.mdata:
+            for f in self.find_log_files('biscuit/%s' % k):
+                d = getattr(self, 'parse_logs_%s' % k)(f['f'], f['fn'])
+                self.mdata[k][f['fn']] = d
+
+        num_parsed = 0
+        for k in self.mdata:
+            num_parsed += len(self.mdata[k])
+        if num_parsed == 0:
+            raise UserWarning
+
+        # Basic Stats Table
+        self.biscuit_stats_table()
+
+        # Write out to the report
+        for k in self.mdata:
+            if len(self.mdata[k]) > 0:
+                self.write_data_file(self.mdata[k], 'multiqc_biscuit_%s' % k)
+                log.info("Found %d biscuit %s reports" % (len(self.mdata[k]), k))
+                if k != 'markdup':
+                    getattr(self, 'chart_biscuit_%s' % k)()
+
+    def biscuit_stats_table(self):
+
+        pd = {}
+        for sid, dd in self.mdata['mapq'].items():
+            if not sid.endswith('_mapq_table.txt'):
+                continue
+            sid = sid.replace('_mapq_table.txt','')
```

Comment:
Please run through the `self.clean_s_name()` function as well. This is needed for the `-b` function and custom sample name cleanups.

See http://multiqc.info/docs/#custom-sample-names

Reply:
Ah, I just got to the search pattern config. Yes, so it's better to have a single search pattern config for every _type_ of file rather than relying on the filename for this reason. Then if each file type is parsed into a separate data structure, you shouldn't need to have this check.

Reply:
I added clean_s_name
```
s_name = self.clean_s_name(f['fn'], f['root']) # allow user to specify directory
self.mdata[k][s_name] = d
```

I did my best to separate search pattern for each file type and removed existing file checking best as I can.